### DISCUSSION TITLE
About async backups

### DISCUSSION THREAD

**Discussion by lewisl9029** [seaweedfs/seaweedfs#3529] [Q&A] [answered]
I just set up async backups to S3 in my dev environment following the instructions here: https://github.com/seaweedfs/seaweedfs/wiki/Async-Backup

A few things I wanted to discuss:

1. Seeing a bunch of logs with the error `completeMultipartUpload resources/JTLtLXSjjlP3nN4tjbNfpCYzrF08EcSVCHbpa8XrUJE: EntityTooSmall: Your proposed upload is smaller than the minimum allowed size` at the end. Full log looks like this:

```
E0827 00:27:41.569969 filer_pb_tail.go:79 process directory:"/resources"  event_notification:{new_entry:{name:"JTLtLXSjjlP3nN4tjbNfpCYzrF08EcSVCHbpa8XrUJE"  chunks:{file_id:"7,226b3fb24ef9"  size:4194304  mtime:1661135791927660085  e_tag:"Bpp6x011IEhwtTl6OWbgew=="  fid:{volume_id:7  file_key:8811  cookie:1068650233}}  chunks:{file_id:"3,226ccbf95a37"  offset:4194304  size:525967  mtime:1661135791855596419  e_tag:"5hzWkEk2zhUhh/SKQaCm7w=="  fid:{volume_id:3  file_key:8812  cookie:3422116407}}  attributes:{file_size:4720271  mtime:1661135791  file_mode:432  crtime:1661135791  mime:"application/json"  md5:">T{[\xde4\xabȍ\n\x8d\x0c\xf6T$\xd7"}  extended:{key:"Seaweed-Meta"  value:"{\"headers\":{\"cache-control\":\"public, max-age=31536000, immutable\",\"content-type\":\"application/json\"}}"}}  delete_chunks:true  new_parent_path:"/resources"  signatures:-2011343073}  ts_ns:1661135791928717585: [reflame-storage-backup-development] completeMultipartUpload resources/JTLtLXSjjlP3nN4tjbNfpCYzrF08EcSVCHbpa8XrUJE: EntityTooSmall: Your proposed upload is smaller than the minimum allowed size
```

Here's a zip with the file in the log: [JTLtLXSjjlP3nN4tjbNfpCYzrF08EcSVCHbpa8XrUJE.zip](https://github.com/seaweedfs/seaweedfs/files/9437048/JTLtLXSjjlP3nN4tjbNfpCYzrF08EcSVCHbpa8XrUJE.zip)

Happy to turn this into an issue if it looks like a valid bug?

2. In my seaweed cluster I'm using a `Seaweed-meta` header to store a bunch of json metadata for each file, but in the s3 backup, I'm not seeing anything in S3 metadata except Content-Type: 

![image](https://user-images.githubusercontent.com/6934200/187008116-f6db44af-ed2a-4d38-90f0-e4fc73c11b0b.png)

Are the Seaweed- headers backed up somewhere in s3 as well?

3. How should I go about testing a restore from the async s3 backup, to make sure it works as expected and document the process for when I need to do this in production? Haven't been able to find any docs around this.

Cheers!

**Comment by chrislusf** [comment, marked as answer]
This is an issue since S3 requires the minimal multipart upload size is 5Mb  http://docs.aws.amazon.com/AmazonS3/latest/API/mpUploadUploadPart.html

**Reply by chrislusf** [reply]
Added a fix for https://github.com/seaweedfs/seaweedfs/issues/3531

**Reply by chrislusf** [reply]
2. headers are stored as tags in AWS S3

**Reply by chrislusf** [reply]
3. There are no one-click to recover. But the backup is a set of normal files. You should be able to recover the lost file by copying.

**Reply by lewisl9029** [reply]
Wow, thank you so much for the quick fix! Will give this another try tomorrow.

Re: 2. I'm not seeing tags for any of the files in the S3 bucket either:
<img width="1318" alt="image" src="https://user-images.githubusercontent.com/6934200/187020316-9dacc323-ff8a-4ba9-969f-604d47f8a253.png">

Also, not sure if related, but I just set up [active-active replication](https://github.com/seaweedfs/seaweedfs/wiki/Filer-Active-Active-cross-cluster-continuous-synchronization) in my local dev environment as well to test it out, and the replicated destination seems to be also missing Seaweed-* headers from the source. 🤔

Will also try to set up a minimal docker-compose based repro for this tomorrow to help with debug.

**Reply by lewisl9029** [reply]
Opened an issue https://github.com/seaweedfs/seaweedfs/issues/3532 along with a repo https://github.com/lewisl9029/seaweed-async-replication-repro to help reproduce.
